include/crypto/event.h

```diff
@@ -0,0 +1,378 @@
+
+    /*
+     * For a multi-priority queue implementation, this function helps
+     * determine in which event stack the event should end up in.
+     *
+     * Expected to return the priority (which may be zero), or -1 on error
+     */
+    int (*determine_priority)(struct ossl_event_st *event);
+
+    /*
+     * Whenever events have been dispatched to the subscribers, the diverse
+     * return values are counted, and this function is used to determined
+     * how to act on that result.  The returned value is simply returned
+     * back to whoever is calling.
+     *
+     * |count_errors| is the count of any subscriber call that returned
+     * |ossl_event_error|.
+     * |count_success| is the count of any subscriber call that returned
+     * |ossl_event_pass| or |ossl_event_handled|.
+     *
+     * Expected to return 1 for good (may continue to be used), -1 for bad
+     * (this queue is not functional any more).
+     */
+    int (*determine_state)(size_t count_errors, size_t count_success);
+
+    /*
+     * Set / get values to / from auxilliary queue method data, or perform
+     * other queue method specific tasks.
+     *
+     * The functionality this implements is open ended, and allows
+     * extraordinary operations such as reordering (rescheduling) events,
+     * affecting other associated queues, as well as modifications or access
+     * of the auxialliary method data, etc etc etc.
+     * Use with care!
+     *
+     * The arguments are of the same style as ioctl.h functions.  The caller
+     * must know exactly what arguments are expected for each |request|.
+     *
+     * Expected to return 1 on success, -1 on failure.
+     */
+    int (*ctrl)(struct ossl_event_queue_st *queue, uint32_t request,
+                ... /* args */);
+    /* va_list variant of the control function */
+    int (*vctrl)(struct ossl_event_queue_st *queue, uint32_t request,
+                 va_list args);
+
+/*
+ * Request the nummber of priorities for this method.
+ * Takes on extra argument:
+ * - a pointer to size_t, where the number of priorities will be stored
+ *
+ * This is a special request; it expects NULL for |queue|, and gives back a
+ * constant number.
+ */
+# define OSSL_EVENT_C_PRIORITIES    0x00000001
+};
+
+/*-
+ * Functions to create and populate events
+ * ---------------------------------------
+ */
+
+/* Allocate / deallocate an event structure dynamically and populate it */
+struct ossl_event_st *ossl_event_new(uint32_t type, void *ctx,
+                                     const void *identifiers,
+                                     void *payload, size_t payload_size,
+                                     ossl_event_destructor_fn *destructor);
+void ossl_event_free(struct ossl_event_st *);
+
+/*
+ * Populate an event structure.  Used for events that aren't allocated
+ * dynamically.
+ *
+ * Returns one of these values:
+ * 0    Not set
+ * 1    Set
+ */
+int ossl_event_set0(struct ossl_event_st *event,
```

Comment:
I'd drop the _0_ in the function name :grin:

Reply:
I think our coding style says that there should be an `0`...  or that one of the recent coding style PRs do.
Anyway, we have use 0/1 to clarify what happens with the data being pointed at for quite a long time.  Why would this be any different?